opentelemetry-sdk/src/opentelemetry/sdk/error_handler/__init__.py

```diff
@@ -0,0 +1,149 @@
+registered in this entry point, this class must inherit from the
+``opentelemetry.sdk.error_handler.ErrorHandler`` class and implement the
+corresponding ``handle`` method. This method will receive the exception object
+that is to be handled. The error handler class should also inherit from the
+exception classes it wants to handle. For example, this would be an error
+handler that handles ``ZeroDivisionError``:
+
+.. code:: python
+
+    from opentelemetry.sdk.error_handler import ErrorHandler
+    from logging import getLogger
+
+    logger = getLogger(__name__)
+
+
+    class ErrorHandler0(ErrorHandler, ZeroDivisionError):
+
+        def _handle(self, error: Exception, *args, **kwargs):
+
+            logger.exception("ErrorHandler0 handling a ZeroDivisionError")
+
+To use the global error handler, just instantiate it as a context manager where
+you want exceptions to be handled:
+
+
+.. code:: python
+
+    from opentelemetry.sdk.error_handler import GlobalErrorHandler
+
+    with GlobalErrorHandler():
+        1 / 0
+
+If the class of the exception raised in the scope of the ``GlobalErrorHandler``
+object is not parent of any registered error handler, then the default error
+handler will handle the exception. This default error handler will only log the
+exception to standard logging, the exception won't be raised any further.
+"""
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+
+from pkg_resources import iter_entry_points
+
+logger = getLogger(__name__)
+
+
+class ErrorHandler(ABC):
+    @abstractmethod
+    def _handle(self, error: Exception, *args, **kwargs):
+        """
+        Handle an exception
+        """
+
+
+class DefaultErrorHandler(ErrorHandler):
+    """
+    Default error handler
+
+    This error handler just logs the exception using standard logging.
+    """
+
+    # pylint: disable=useless-return
+    def _handle(self, error: Exception, *args, **kwargs):
+
+        logger.exception("Error handled by default error handler: ")
+        return None
+
+
+class GlobalErrorHandler:
+    """
+    Global error handler
+
+    This is a singleton class that can be instantiated anywhere to get the
+    global error handler. This object provides a handle method that receives
+    an exception object that will be handled by the registered error handlers.
+    """
+
+    _instance = None
+
+    def __new__(cls) -> "GlobalErrorHandler":
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+
+        return cls._instance
+
+    def __enter__(self):
+        pass
+
+    # pylint: disable=no-self-use
+    def __exit__(self, exc_type, exc_value, traceback):
+
+        if exc_value is None:
+
+            return None
+
+        plugin_handled = False
+
+        for error_handler_entry_point in iter_entry_points(
+            "opentelemetry_error_handler"
+        ):
+
+            error_handler_class = error_handler_entry_point.load()
```

Comment:
I don't think I can see anything that enforces a specific handler to handle only certain types of errors (inheriting for an `Error` doesn't really enforce this). With that being said, this logic will run through all the handlers regardless of what kind of errors they are handling. Does this make sense if there are multiple handlers handling the same type of error already?

Reply:
I imagine most implementations of the handlers having to deal with checking the exception type if this is not implemented somewhere else. It seems natural to me that handlers are made for specific exceptions, and they will not be interested in any other exception types, so I preferred to put this check in the global error handler. I am open to other design ideas, this is not something I consider definitive for this particular feature.

Reply:
That's fine then. I think we can leave it up to the user's discretion to check exception type and handle them appropriately (if they have multiple handlers executing on the same error type then that is up to them). Is the order of execution of the handlers known somehow? Or does that depend on which iter point is loaded first?

Reply:
That would depend on the `iter_points` order. So, you prefer that the global error handler does not check exception type and all error handlers are called?

Reply:
I think the current implementation is fine.